File: C/Computer Systems/bitwise/swap.c
#include <stdio.h>

void swapper(int *x, int *y);
void reverse_array(int a[], int cnt);
void printarr(int arr[], int len);

int main(void) {
  int arr[] = {1, 2, 3, 4, 5, 6, 7};

  printarr(arr, 7);

  reverse_array(arr, 7);

  printarr(arr, 7);

  return 0;
}

void swapper(int *x, int *y) {
  *y = *x ^ *y;
  *x = *x ^ *y;
  *y = *x ^ *y;
}

void reverse_array(int a[], int cnt) {
  int *first, *last;
  for (first = a, last = a + cnt - 1; first < last; first++, last--) {
    swapper(first, last);
  }
}

void printarr(int arr[], int len) {
  printf("[");
  for (int i = 0; i < len; i++) {
    printf(" %d,", arr[i]);
  }
  printf("]\n");
}

File: Rust/Computer Systems/computer_systems/src/bitwise/swap.rs
/// # Safety
/// just be sure the lenght of the array is greater or equal to the parameter len
pub unsafe fn printarr(arr: *const i32, len: usize) {
    print!("[");
    for i in 0..len {
        unsafe { print!(" {},", *arr.add(i)) }
    }
    println!(" ]");
}

#[allow(clippy::manual_swap)]
fn swapper(x: &mut i32, y: &mut i32) {
    *y ^= *x;
    *x ^= *y;
    *y ^= *x;
}

pub fn reverse_array<const N: usize>(arr: &mut [i32; N]) {
    let mid = N / 2;
    let (left, right) = arr.split_at_mut(mid);
    for i in 0..mid {
        swapper(&mut left[i], &mut right[mid - i]);
    }
}
